app/controllers/api/avatars.php

```diff
@@ -49,11 +53,92 @@
+        $accessTokenExpiry = $session->getAttribute('providerAccessTokenExpiry');
+        $refreshToken = $session->getAttribute('providerRefreshToken');
+
+        $appId = $project->getAttribute('authProviders', [])[$provider . 'Appid'] ?? '';
+        $appSecret = $project->getAttribute('authProviders', [])[$provider . 'Secret'] ?? '{}';
+
+        $className = 'Appwrite\\Auth\\OAuth2\\' . \ucfirst($provider);
+
+        if (!\class_exists($className)) {
+            throw new Exception(Exception::PROJECT_PROVIDER_UNSUPPORTED);
+        }
+
+        $oauth2 = new $className($appId, $appSecret, '', [], []);
+
+        $isExpired = new \DateTime($accessTokenExpiry) < new \DateTime('now');
+        if ($isExpired) {
+            try {
+                $oauth2->refreshTokens($refreshToken);
+
+                $accessToken = $oauth2->getAccessToken('');
+                $refreshToken = $oauth2->getRefreshToken('');
+
+                $verificationId = $oauth2->getUserID($accessToken);
+
+                if (empty($verificationId)) {
+                    throw new \Exception("Locked tokens."); // Race codition, handeled in catch
+                }
+
+                $session
+                    ->setAttribute('providerAccessToken', $accessToken)
+                    ->setAttribute('providerRefreshToken', $refreshToken)
+                    ->setAttribute('providerAccessTokenExpiry', DateTime::addSeconds(new \DateTime(), (int)$oauth2->getAccessTokenExpiry('')));
+
+                Authorization::skip(fn () => $dbForProject->updateDocument('sessions', $session->getId(), $session));
+
+                $dbForProject->deleteCachedDocument('users', $user->getId());
+            } catch (Throwable $err) {
+                $index = 0;
+                do {
+                    $previousAccessToken = $session->getAttribute('providerAccessToken');
+
+                    $user = Authorization::skip(fn () => $dbForConsole->getDocument('users', $userId));
+                    $sessions = $user->getAttribute('sessions', []);
+                    $session = $sessions[0] ?? new Document();
+                    $accessToken = $session->getAttribute('providerAccessToken');
+
+                    if ($accessToken !== $previousAccessToken) {
+                        break;
+                    }
+
+                    $index++;
+                    \usleep(500000);
+                } while ($index < 10);
+            }
+        }
+
+        $oauth2 = new $className($appId, $appSecret, '', [], []);
+        $githubUser = $oauth2->getUserSlug($accessToken);
+        $githubId = $oauth2->getUserID($accessToken);
+
+        return [
+            'name' => $githubUser,
+            'id' => $githubId
+        ];
+    } catch (Exception $err) {
+        \var_dump($err->getMessage());
+        \var_dump($err->getTraceAsString());
+        \var_dump($err->getLine());
+        \var_dump($err->getFile());
+        return [];
+    }
```

Comment:
You could get rid of this catch block and replace it with a `finally` block

Reply:
I wanted to print logs. If someone complains, we can check what's wrong on cloud. Maybe there are some GitHub profile privacy options that would break something. maybe.

Reply:
You can just throw a 500 error and it will report it to sentry